Updated for Julia 1.0, added (limited) auto detection of libraries and avx support (#17)
JuliaMath/VML.jl: README.md

# VML

This package provides bindings to the Intel Vector Math Library for
arithmetic and transcendental functions. Especially for large vectors it is often substantially faster than broadcasting Julia's built-in functions.

## Setting up VML.jl

To use VML.jl, you must have the Intel Vector Math Library installed.
For this you have two options. First, you can install (and build) [MKL.jl](https://github.com/JuliaComputing/MKL.jl), which will add the necessary libraries to your Julia install. This will change your Julia system image however, so if you would prefer not to do that you can get the stand-alone [MKL](http://software.intel.com/en-us/intel-mkl),

> which is free for non-commercial use. 

Reviewer: As pointed out in this PR (https://github.com/JuliaMath/VML.jl/pull/16), Intel allows any kind of usage: https://software.intel.com/en-us/mkl/license-faq, as long as the library is attachable by the user

Author: Done

> 

Reviewer: To use MKL.jl and then VML.jl:
```julia
julia>]add https://github.com/JuliaComputing/MKL.jl

julia>]add https://github.com/JuliaMath/VML.jl
```
After restarting the REPL:
```
julia> using VML
```

Libdl automatically finds the relevant libraries in the default search locations. This already the case if you use MKL.jl, but the stand-alone may require you to source `mklvars.sh`. The default command on Mac and Ubuntu is `source /opt/intel/mkl/bin/mklvars.sh intel64`. You may want to add this to your `.bashrc`. 
Adding a new `*.conf` file in `/etc/ld.so.conf.d` also works, as the `intel-mkl-slim` package in the AUR does automatically. 

> Using [CpuId.jl](https://github.com/m-j-w/CpuId.jl), VML.jl detects if your processor supports the newer `avx2` instructions, and if not default to `libmkl_vml_avx`. If your system does not have AVX this package will currently not work for you.

Reviewer: These should be moved to a new section called

## Developing

Normal user doesn't need to know about what happens inside VML.

Author: I see your point. I have updating the Benchmarks and with it the Readme on my ToDo list, but did not get to it. I might have some time for it tomorrow.
At the same time, some it is Installation Instructions, but these could possibly be split into Basic and Advanced.

Reviewer: Awesome! I added some other comments on Readme that you can use directly instantly. ;)

Author: Done

If the CPU feature detection does not work for you, please open an issue. 

## Using VML.jl

After loading VML.jl, vector calls to functions listed below will

> automatically use VML instead of openlibm when possible. Note that most function currently do not have a vectorized version (e.g. you call `sin.(rand(300))`), so there should be no conflict. If there is let me know. Updated and conflict tested exported functions are planned for the future.

Reviewer 2: There will be conflicts for `exp` and `log`, which act on matrices in Base, as do many trig functions. 
```
julia> exp([1 1; 1 1.0])
2×2 Array{Float64,2}:
 4.19453  3.19453
 3.19453  4.19453

julia> VML.exp([1 1; 1 1.0])
2×2 Array{Float64,2}:
 2.71828  2.71828
 2.71828  2.71828
```

Author: Added the example to the readme

# VML

This package provides bindings to the Intel Vector Math Library for
arithmetic and transcendental functions. Especially for large vectors it is often substantially faster than broadcasting Julia's built-in functions.

## Setting up VML.jl

To use VML.jl, you must have the Intel Vector Math Library installed.
For this you have two options. First, you can install (and build) [MKL.jl](https://github.com/JuliaComputing/MKL.jl), which will add the necessary libraries to your Julia install. This will change your Julia system image however, so if you would prefer not to do that you can get the stand-alone [MKL](http://software.intel.com/en-us/intel-mkl),
which is free for non-commercial use. The default install location will be detected automatically (currently only `opt/intel/mkl/lib` on Unix). If you have chosen a different location or have multiple versions of MKL, you can also specify the environmental variable 
```

>     ENV["MKL_SL"] = <.../lib>

Reviewer 3: Why not use `MKL_ROOT` for this which is what you get when you run the MKL provided script (https://software.intel.com/en-us/mkl-linux-developer-guide-scripts-to-set-environment-variables)

Author: When I installed MKL the MKLROOT variable was not added automatically, instead I had to add the script to my `.*rc` manually.
( Apt-get install on Ubuntu, and AUR install on arch) 
On another machine I deliberately did not add it, as the PATH was already quite long.
So I decided to just check for the default path as a fallback. 
But I agree, I should add a check for MKLROOT as well.

Thank you for reviewing this.

Reviewer 3: But you need to set the `MKL_ROOT` to use MKL in most situations (e.g. when linking to it). To me, it seems nice to use the same way to detect MKL as everyone else is using instead of introducing our own, package specific environment variable.

Author: Ok, I have learned quite a bit about dynamic libaries and Libdl since I wrote this, and I have amended my commit to leave out all that stuff. 
In my quest to be as general as possible I missed that it is not really necessary. 
I think is is now basically ready.

```
The specified folder should contain `libmkl_rt`, `libmkl_core` and `libmkl_vml_avx` or `libmkl_vml_avx2`, with file endings appropriate for your operating system. 
The definition of `ENV["MKL_SL"]` will take precedence even if `MKL.jl` or default standalone MKL are installed. 

> Using [CpuId.jl](https://github.com/m-j-w/CpuId.jl) VML.jl detects if your processor supports the newer `avx2` instructions, and if not default to `libmkl_vml_avx`. If your system does not have AVX this package will currently not work for you.

Reviewer 3: ```suggestion
Using [CpuId.jl](https://github.com/m-j-w/CpuId.jl), VML.jl detects if your processor supports the newer `avx2` instructions, and if not default to `libmkl_vml_avx`. If your system does not have AVX this package will currently not work for you.
```

> If any of this detection does not work for you, please let me know. 

Reviewer 3: ```suggestion
If the CPU feature detection does not work for you, please open an issue. 
```

[Intel's website](http://software.intel.com/sites/products/documentation/hpc/mkl/vml/vmldata.htm).

> 

Reviewer: # Example usage
```
julia> using VML
julia> a = rand(10000);
julia>@time  sin.(a);                                                                                                    
 0.159878 seconds (583.25 k allocations: 30.720 MiB, 2.78% gc time)
julia> @time sin(a);                                                                                                      
0.000465 seconds (6 allocations: 781.484 KiB) 
```

Reviewer 2: Note that this example will change after  https://github.com/Crown421/VML.jl/pull/1 , to need an explicit `VML.sin(a)`.

Author: Changed

### Unary functions

Allocating forms have signature `f(A)`. Mutating forms have signatures

> `f!(A)` (in place) and `f!(out, A)` (out of place). The last 9 functions have been moved from Base to `SpecialFunctions.jl` or have not Base equivalent. 

Reviewer 3: ```suggestion
`f!(A)` (in place) and `f!(out, A)` (out of place). The last 9 functions have been moved from Base to `SpecialFunctions.jl` or have no Base equivalent. 
```

### Binary functions

Allocating forms have signature `f(A, B)`. Mutating forms have
signature `f!(out, A, B)`. 

Allocating | Mutating
-----------|---------
`atan`    | `atan!`
`hypot`    | `hypot!`
`pow`       | `pow!`
`divide`       | `divide!`

### Next steps

> Next steps for this package are proper Windows support, writing up proper testing to make sure each advertised function actually works as expected and build testing on travis.

Reviewer: ```suggestion
Next steps for this package are writing up proper testing to make sure each advertised function actually works as expected and build testing on Travis.
```

Author: Added
